File: android/src/main/java/io/split/splitio/SplitClientConfigHelper.java
package io.split.splitio;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.split.android.client.ServiceEndpoints;
import io.split.android.client.SplitClientConfig;
import io.split.android.client.SplitFilter;
import io.split.android.client.SyncConfig;
import io.split.android.client.impressions.ImpressionListener;

class SplitClientConfigHelper {

    private static final String FEATURES_REFRESH_RATE = "featuresRefreshRate";
    private static final String SEGMENTS_REFRESH_RATE = "segmentsRefreshRate";
    private static final String IMPRESSIONS_REFRESH_RATE = "impressionsRefreshRate";
    private static final String TELEMETRY_REFRESH_RATE = "telemetryRefreshRate";
    private static final String EVENTS_QUEUE_SIZE = "eventsQueueSize";
    private static final String IMPRESSIONS_QUEUE_SIZE = "impressionsQueueSize";
    private static final String EVENT_FLUSH_INTERVAL = "eventFlushInterval";
    private static final String EVENTS_PER_PUSH = "eventsPerPush";
    private static final String TRAFFIC_TYPE = "trafficType";
    private static final String ENABLE_DEBUG = "enableDebug";
    private static final String STREAMING_ENABLED = "streamingEnabled";
    private static final String PERSISTENT_ATTRIBUTES_ENABLED = "persistentAttributesEnabled";
    private static final String SDK_ENDPOINT = "sdkEndpoint";
    private static final String EVENTS_ENDPOINT = "eventsEndpoint";
    private static final String SSE_AUTH_SERVICE_ENDPOINT = "authServiceEndpoint";
    private static final String STREAMING_SERVICE_ENDPOINT = "streamingServiceEndpoint";
    private static final String TELEMETRY_SERVICE_ENDPOINT = "telemetryServiceEndpoint";
    private static final String IMPRESSION_LISTENER = "impressionListener";
    private static final String SYNC_CONFIG = "syncConfig";
    private static final String SYNC_CONFIG_NAMES = "syncConfigNames";
    private static final String SYNC_CONFIG_PREFIXES = "syncConfigPrefixes";

    /**
     * Creates a {@link SplitClientConfig} object from a map.
     *
     * @param configurationMap Map of config values.
     * @param impressionListener Optional ImpressionListener.
     * @return {@link SplitClientConfig} object.
     */
    static SplitClientConfig fromMap(@NonNull Map<String, Object> configurationMap, @Nullable ImpressionListener impressionListener) {
        SplitClientConfig.Builder builder = SplitClientConfig.builder();

        Integer featuresRefreshRate = getInteger(configurationMap, FEATURES_REFRESH_RATE);
        if (featuresRefreshRate != null) {
            builder.featuresRefreshRate(featuresRefreshRate);
        }

        Integer segmentsRefreshRate = getInteger(configurationMap, SEGMENTS_REFRESH_RATE);
        if (segmentsRefreshRate != null) {
            builder.segmentsRefreshRate(segmentsRefreshRate);
        }

        Integer impressionsRefreshRate = getInteger(configurationMap, IMPRESSIONS_REFRESH_RATE);
        if (impressionsRefreshRate != null) {
            builder.impressionsRefreshRate(impressionsRefreshRate);
        }

        Integer telemetryRefreshRate = getInteger(configurationMap, TELEMETRY_REFRESH_RATE);
        if (telemetryRefreshRate != null) {
            builder.telemetryRefreshRate(telemetryRefreshRate);
        }

        Integer eventsQueueSize = getInteger(configurationMap, EVENTS_QUEUE_SIZE);
        if (eventsQueueSize != null) {
            builder.eventsQueueSize(eventsQueueSize);
        }

        Integer impressionsQueueSize = getInteger(configurationMap, IMPRESSIONS_QUEUE_SIZE);
        if (impressionsQueueSize != null) {
            builder.impressionsQueueSize(impressionsQueueSize);
        }

        Integer eventFlushInterval = getInteger(configurationMap, EVENT_FLUSH_INTERVAL);
        if (eventFlushInterval != null) {
            builder.eventFlushInterval(eventFlushInterval);
        }

        Integer eventsPerPush = getInteger(configurationMap, EVENTS_PER_PUSH);
        if (eventsPerPush != null) {
            builder.eventsPerPush(eventsPerPush);
        }

        String trafficType = getString(configurationMap, TRAFFIC_TYPE);
        if (trafficType != null) {
            builder.trafficType(trafficType);
        }

        Boolean enableDebug = getBoolean(configurationMap, ENABLE_DEBUG);
        if (enableDebug != null && enableDebug) {
            builder.enableDebug();
        }

        Boolean streamingEnabled = getBoolean(configurationMap, STREAMING_ENABLED);
        if (streamingEnabled != null) {
            builder.streamingEnabled(streamingEnabled);
        }

        Boolean persistentAttributesEnabled = getBoolean(configurationMap, PERSISTENT_ATTRIBUTES_ENABLED);
        if (persistentAttributesEnabled != null) {
            builder.persistentAttributesEnabled(persistentAttributesEnabled);
        }

        ServiceEndpoints.Builder serviceEndpointsBuilder = ServiceEndpoints
                .builder();

        String apiEndpoint = getString(configurationMap, SDK_ENDPOINT);
        if (apiEndpoint != null) {
            serviceEndpointsBuilder.apiEndpoint(apiEndpoint);
        }

        String eventsEndpoint = getString(configurationMap, EVENTS_ENDPOINT);
        if (eventsEndpoint != null) {
            serviceEndpointsBuilder.eventsEndpoint(eventsEndpoint);
        }

        String sseAuthServiceEndpoint = getString(configurationMap, SSE_AUTH_SERVICE_ENDPOINT);
        if (sseAuthServiceEndpoint != null) {
            serviceEndpointsBuilder.sseAuthServiceEndpoint(sseAuthServiceEndpoint);
        }

        String streamingServiceEndpoint = getString(configurationMap, STREAMING_SERVICE_ENDPOINT);
        if (streamingServiceEndpoint != null) {
            serviceEndpointsBuilder.streamingServiceEndpoint(streamingServiceEndpoint);
        }

        String telemetryServiceEndpoint = getString(configurationMap, TELEMETRY_SERVICE_ENDPOINT);
        if (telemetryServiceEndpoint != null) {
            serviceEndpointsBuilder.telemetryServiceEndpoint(telemetryServiceEndpoint);
        }

        if (impressionListener != null) {
            builder.impressionListener(impressionListener);
        }

        Map<String, List<String>> syncConfig = getListMap(configurationMap, SYNC_CONFIG);
        if (syncConfig != null) {
            List<String> names = syncConfig.get(SYNC_CONFIG_NAMES);
            List<String> prefixes = syncConfig.get(SYNC_CONFIG_PREFIXES);

            SyncConfig.Builder syncConfigBuilder = SyncConfig.builder();
            if (names != null && !names.isEmpty()) {
                syncConfigBuilder.addSplitFilter(SplitFilter.byName(names));
            }

            if (prefixes != null && !prefixes.isEmpty()) {
                syncConfigBuilder.addSplitFilter(SplitFilter.byPrefix(prefixes));
            }

            builder.syncConfig(syncConfigBuilder.build());
        }

        return builder.serviceEndpoints(serviceEndpointsBuilder.build()).build();
    }

    static boolean impressionListenerEnabled(@NonNull Map<String, Object> configurationMap) {
        Boolean impressionListenerEnabled = getBoolean(configurationMap, IMPRESSION_LISTENER);
        return impressionListenerEnabled != null && impressionListenerEnabled;
    }

    @Nullable
    private static Integer getInteger(Map<String, Object> map, String key) {
        if (map.containsKey(key)) {
            Object value = map.get(key);
            if (value != null && value.getClass().isAssignableFrom(Integer.class)) {
                return (Integer) value;
            }
        }

        return null;
    }

    @Nullable
    private static String getString(Map<String, Object> map, String key) {
        if (map.containsKey(key)) {
            Object value = map.get(key);
            if (value != null && value.getClass().isAssignableFrom(String.class)) {
                return (String) value;
            }
        }

        return null;
    }

    @Nullable
    private static Boolean getBoolean(Map<String, Object> map, String key) {
        if (map.containsKey(key)) {
            Object value = map.get(key);
            if (value != null && value.getClass().isAssignableFrom(Boolean.class)) {
                return (Boolean) value;
            }
        }

        return null;
    }

    @Nullable
    private static Map<String, List<String>> getListMap(Map<String, Object> map, String key) {
        if (map.containsKey(key)) {
            Object value = map.get(key);
            if (value != null && value.getClass().isAssignableFrom(HashMap.class)) {
                return (HashMap<String, List<String>>) value;
            }
        }

        return null;
    }
}

File: splitio_ios/ios/Classes/SplitClientConfigHelper.swift
import Foundation
import Split

class SplitClientConfigHelper {

    static private let FEATURES_REFRESH_RATE = "featuresRefreshRate"
    static private let SEGMENTS_REFRESH_RATE = "segmentsRefreshRate"
    static private let IMPRESSIONS_REFRESH_RATE = "impressionsRefreshRate"
    static private let TELEMETRY_REFRESH_RATE = "telemetryRefreshRate"
    static private let EVENTS_QUEUE_SIZE = "eventsQueueSize"
    static private let IMPRESSIONS_QUEUE_SIZE = "impressionsQueueSize"
    static private let EVENT_FLUSH_INTERVAL = "eventFlushInterval"
    static private let EVENTS_PER_PUSH = "eventsPerPush"
    static private let TRAFFIC_TYPE = "trafficType"
    static private let ENABLE_DEBUG = "enableDebug"
    static private let STREAMING_ENABLED = "streamingEnabled"
    static private let PERSISTENT_ATTRIBUTES_ENABLED = "persistentAttributesEnabled"
    static private let SDK_ENDPOINT = "sdkEndpoint"
    static private let EVENTS_ENDPOINT = "eventsEndpoint"
    static private let SSE_AUTH_SERVICE_ENDPOINT = "authServiceEndpoint"
    static private let STREAMING_SERVICE_ENDPOINT = "streamingServiceEndpoint"
    static private let TELEMETRY_SERVICE_ENDPOINT = "telemetryServiceEndpoint"
    static private let IMPRESSION_LISTENER = "impressionListener"
    static private let SYNC_CONFIG = "syncConfig"
    static private let SYNC_CONFIG_NAMES = "syncConfigNames"
    static private let SYNC_CONFIG_PREFIXES = "syncConfigPrefixes"
    static private let SYNC_CONFIG_SETS = "syncConfigFlagSets"
    static private let IMPRESSIONS_MODE = "impressionsMode"
    static private let SYNC_ENABLED = "syncEnabled"
    static private let USER_CONSENT = "userConsent"
    static private let ENCRYPTION_ENABLED = "encryptionEnabled"
    static private let LOG_LEVEL = "logLevel"
    static private let READY_TIMEOUT = "readyTimeout"

    static func fromMap(configurationMap: [String: Any?], impressionListener: SplitImpressionListener?) -> SplitClientConfig {
        let config = SplitClientConfig()

        if configurationMap[FEATURES_REFRESH_RATE] != nil {
            if let featuresRefreshRate = configurationMap[FEATURES_REFRESH_RATE] as? Int {
                config.featuresRefreshRate = featuresRefreshRate
            }
        }

        if configurationMap[SEGMENTS_REFRESH_RATE] != nil {
            if let segmentsRefreshRate = configurationMap[SEGMENTS_REFRESH_RATE] as? Int {
                config.segmentsRefreshRate = segmentsRefreshRate
            }
        }

        if configurationMap[IMPRESSIONS_REFRESH_RATE] != nil {
            if let impressionsRefreshRate = configurationMap[IMPRESSIONS_REFRESH_RATE] as? Int {
                config.impressionRefreshRate = impressionsRefreshRate
            }
        }

        if configurationMap[TELEMETRY_REFRESH_RATE] != nil {
            if let telemetryRefreshRate = configurationMap[TELEMETRY_REFRESH_RATE] as? Int {
                config.telemetryRefreshRate = telemetryRefreshRate
            }
        }

        if configurationMap[EVENTS_QUEUE_SIZE] != nil {
            if let eventsQueueSize = configurationMap[EVENTS_QUEUE_SIZE] as? Int {
                config.eventsQueueSize = Int64(eventsQueueSize)
            }
        }

        if configurationMap[IMPRESSIONS_QUEUE_SIZE] != nil {
            if let impressionsQueueSize = configurationMap[IMPRESSIONS_QUEUE_SIZE] as? Int {
                config.impressionsQueueSize = impressionsQueueSize
            }
        }

        if configurationMap[EVENT_FLUSH_INTERVAL] != nil {
            if let eventFlushInterval = configurationMap[EVENT_FLUSH_INTERVAL] as? Int {
                config.eventsPushRate = eventFlushInterval
            }
        }

        if configurationMap[EVENTS_PER_PUSH] != nil {
            if let eventsPerPush = configurationMap[EVENTS_PER_PUSH] as? Int {
                config.eventsPerPush = eventsPerPush
            }
        }

        if configurationMap[TRAFFIC_TYPE] != nil {
            if let trafficType = configurationMap[TRAFFIC_TYPE] as? String {
                config.trafficType = trafficType
            }
        }

        if configurationMap[ENABLE_DEBUG] != nil {
            if let enableDebug = configurationMap[ENABLE_DEBUG] as? Bool {
                if (enableDebug) {
                    config.logLevel = .debug
                } else {
                    config.logLevel = .none
                }
            }
        }

        if configurationMap[STREAMING_ENABLED] != nil {
            if let streamingEnabled = configurationMap[STREAMING_ENABLED] as? Bool {
                config.streamingEnabled = streamingEnabled
            }
        }

        if configurationMap[PERSISTENT_ATTRIBUTES_ENABLED] != nil {
            if let persistentAttributesEnabled = configurationMap[PERSISTENT_ATTRIBUTES_ENABLED] as? Bool {
                config.persistentAttributesEnabled = persistentAttributesEnabled
            }
        }

        let serviceEndpointsBuilder = ServiceEndpoints.builder()

        if configurationMap[SDK_ENDPOINT] != nil {
            if let sdkEndpoint = configurationMap[SDK_ENDPOINT] as? String {
                serviceEndpointsBuilder.set(sdkEndpoint: sdkEndpoint)
            }
        }

        if configurationMap[EVENTS_ENDPOINT] != nil {
            if let eventsEndpoint = configurationMap[EVENTS_ENDPOINT] as? String {
                serviceEndpointsBuilder.set(eventsEndpoint: eventsEndpoint)
            }
        }

        if configurationMap[SSE_AUTH_SERVICE_ENDPOINT] != nil {
            if let sseAuthServiceEndpoint = configurationMap[SSE_AUTH_SERVICE_ENDPOINT] as? String {
                serviceEndpointsBuilder.set(authServiceEndpoint: sseAuthServiceEndpoint)
            }
        }

        if configurationMap[STREAMING_SERVICE_ENDPOINT] != nil {
            if let streamingServiceEndpoint = configurationMap[STREAMING_SERVICE_ENDPOINT] as? String {
                serviceEndpointsBuilder.set(streamingServiceEndpoint: streamingServiceEndpoint)
            }
        }

        if configurationMap[TELEMETRY_SERVICE_ENDPOINT] != nil {
            if let telemetryServiceEndpoint = configurationMap[TELEMETRY_SERVICE_ENDPOINT] as? String {
                serviceEndpointsBuilder.set(telemetryServiceEndpoint: telemetryServiceEndpoint)
            }
        }

        if let impressionListener = impressionListener {
            config.impressionListener = impressionListener
        }

        if configurationMap[SYNC_CONFIG] != nil {
            if let syncConfig = configurationMap[SYNC_CONFIG] as? [String: [String]] {
                let syncConfigBuilder = SyncConfig.builder()

                if (syncConfig[SYNC_CONFIG_SETS] != nil && syncConfig[SYNC_CONFIG_SETS]?.isEmpty == false) {
                    if let syncFlagSets = syncConfig[SYNC_CONFIG_SETS] as? [String] {
                        syncConfigBuilder.addSplitFilter(SplitFilter.bySet(syncFlagSets))
                    }
                } else {
                    if let syncNames = syncConfig[SYNC_CONFIG_NAMES] as? [String] {
                        syncConfigBuilder.addSplitFilter(SplitFilter.byName(syncNames))
                    }

                    if let syncPrefixes = syncConfig[SYNC_CONFIG_PREFIXES] as? [String] {
                        syncConfigBuilder.addSplitFilter(SplitFilter.byPrefix(syncPrefixes))
                    }
                }

                config.sync = syncConfigBuilder.build()
            }
        }
        
        if let impressionsMode = configurationMap[IMPRESSIONS_MODE] as? String {
            config.impressionsMode = impressionsMode.uppercased()
        }
        
        if let syncEnabled = configurationMap[SYNC_ENABLED] as? Bool {
            config.syncEnabled = syncEnabled
        }

        if let userConsent = configurationMap[USER_CONSENT] as? String {
            switch userConsent.lowercased() {
            case "unknown":
                config.userConsent = .unknown
            case "declined":
                config.userConsent = .declined
            default:
                config.userConsent = .granted
            }
        }

        if let encryptionEnabled = configurationMap[ENCRYPTION_ENABLED] as? Bool {
            config.encryptionEnabled = encryptionEnabled
        }

        if let logLevel = configurationMap[LOG_LEVEL] as? String {
            switch logLevel.lowercased() {
                case "verbose":
                    config.logLevel = .verbose
                case "debug":
                    config.logLevel = .debug
                case "info":
                    config.logLevel = .info
                case "warning":
                    config.logLevel = .warning
                case "error":
                    config.logLevel = .error
                default:
                    config.logLevel = .none
            }
        }

        if let readyTimeout = configurationMap[READY_TIMEOUT] as? Int {
            config.sdkReadyTimeOut = readyTimeout * 1000 // iOS SDK uses this parameter in millis
        }

        config.serviceEndpoints = serviceEndpointsBuilder.build()

        return config
    }

    static func impressionListenerEnabled(configurationMap: [String: Any?]) -> Bool {
        if configurationMap[IMPRESSION_LISTENER] != nil {
            if let impressionListenerEnabled = configurationMap[IMPRESSION_LISTENER] as? Bool {
                return true
            }
        }

        return false
    }
}
